### ISSUE TITLE
[Feature request] Add a possibility to persist artifacts besides the model itself

### ISSUE THREAD

**Issue by benjamin-work** [ottogroup/palladium#46]
At the moment, only the model can be persisted and loaded. However, there are scenarios that necessitate saving and loading additional data.

E.g., assume that we have a regression problem. We want to normalize the targets to a certain range during training but when calling the predict service, data should be mapped back to the original range. Touching the targets is not part of an `sklearn` pipeline, so we may do it during data loading. However, when we start the prediction service, we need to have access to the mapping. Currently, we would have to load the data again to generate the mapping, or try to save the mapping as an attribute of the model.

Ideally, we would be able to just save and load the mapping using palladium tools. The solution should not be too specific to the example above, but be a more general solution to how to persist additional artifacts.

**Comment by dnouri** [comment]
Another way to deal with this is to move the normalization into a model wrapper (or "meta-estimator" in scikit-learn).  A `NormalizeTarget` wrapper would normalize on the way in and out.  The model is somewhat more self-contained this way, which may be good regardless.

**Comment by benjamin-work** [comment]
Yes, for this specific case, that would work. For other cases, that could be an awkward solution. I could imagine that a more general solution would have a "cache" that is just stored together with the model, so that there is no need for handling separate files.

**Comment by dnouri** [comment]
There's this utility called `palladium.interfaces.annotate` which is used by Palladium to store the model version along with the model pickle.  It's a glorified way of sticking an attribute onto the object before it's pickled.

To stick something in you would call `annotate(model, {'useful': 'stuffs'})`, and to get it out again (say in production, after loading): `stuffs = annotate(model)['useful']`.

**Comment by benjamin-work** [comment]
Okay, so you would suggest to use this if extra data needs to be saved?

**Comment by dnouri** [comment]
> Okay, so you would suggest to use this if extra data needs to be saved?

Hmm, just had another look and it seems that at least `palladium.persistence.Database` assumes it can call `json.dumps` on the annotations.  (It then stores the metadata in a separate column.)  So this won't work for all types of data.

Which leaves us with what you already did I assume, which is sticking attributes on the model object.  Not too nice, but probably nicer than having to worry about storing extra data somewhere else and having to support that in all persisters.

If you prefer to use something like `annotate`, then we could make a trivial change and add other keys, besides `__metadata__`, to `annotate`.  (`__metadata__` is what it's trying to be clever about when persisting.)

**Comment by benjamin-work** [comment]
But isn't the model just a blob? Instead of persisting the model, could we not persist something like `{'model': model, 'cache': cache}`? That way, we don't need to store something extra and worry about keeping model and extra in sync.

**Comment by orgesleka** [comment]
@benjamin-work maybe you want to have a look at https://github.com/orgesleka/webscikit which has the feature you are looking for. It stores additional metadata at `model.metada`, which is just a dicitonary you can fill with additional data. The model itself is also stored in this metadatata. For example: If you have a `RandomForestRegressor rfr` you could store it as `metadata['rfr'] = rfr` and then later read it back at `predict(new_data)` or `transform(new_data)` with `rfr = self.metadata['rfr']`
You could also store additional metrics, or `pandas.DataFrames` if you want in `metadata` which would be usefull for later to `predict(new_data)`.

Kind regards
Orges Leka
